daemon/config: fix filter type in BuildKit GC config

tiborvass opened the pull request:
Signed-off-by: Tibor Vass <tibor@docker.com>

Fixes #39601

Commit:
daemon/config: fix filter type in BuildKit GC config

For backwards compatibility, the old incorrect object format for
builder.GC.Rule.Filter still works but is deprecated in favor of array of
strings akin to what needs to be passed on the CLI.

Signed-off-by: Tibor Vass <tibor@docker.com>

cpuguy83 reviewed (approved):
LGTM

tonistiigi commented:
We could possibly use `RawMessage`(for some time) here so we don't break people who used the old format. I guess not a lot of people are affected but if they are it is a failure that causes daemon not to start.

tiborvass commented:
Ok will update to prevent breaking people

thaJeztah commented:
> Ok will update to prevent breaking people

Perhaps as a separate commit, and with a `TODO` added to remove the commit after X release(s) ?

tiborvass commented:
PTAL @tonistiigi

tonistiigi commented:
@tiborvass CI red

tiborvass commented:
@tonistiigi updated

Commit:
daemon/config: add MarshalJSON for future proofing

If anything marshals the daemon config now or in the future
this commit ensures the correct canonical form for the builder
GC policies' filters.

Signed-off-by: Tibor Vass <tibor@docker.com>

tiborvass commented:
@tonistiigi Updated to your liking, and added another commit for safety. PTAL

thaJeztah commented on daemon/config/builder_test.go:
```diff
@@ -0,0 +1,44 @@
+      "policy": [
+        {"keepStorage": "10GB", "filter": ["unused-for=2200h"]},
+        {"keepStorage": "50GB", "filter": {"unused-for": {"3300h": true}}},
+        {"keepStorage": "100GB", "all": true}
```
(just a comment); trying to get my head around how the combination of these policies work 🤔😁

Guess I need to read the docs
